xds/internal/balancer/clusterimpl/clusterimpl.go

```diff
@@ -0,0 +1,313 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package clusterimpl implements the xds_cluster_impl balancing policy. It
+// handles the cluster features (e.g. circuit_breaking, RPC dropping).
+//
+// Note that it doesn't handle name resolution, which is done by policy
+// xds_cluster_resolver.
+package clusterimpl
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/internal/buffer"
+	"google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/serviceconfig"
+	"google.golang.org/grpc/xds/internal/balancer/loadstore"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+	"google.golang.org/grpc/xds/internal/client/load"
+)
+
+const (
+	clusterImplName = "xds_cluster_impl_experimental"
+	// TODO: define defaultRequestCountMax = 1024
+)
+
+func init() {
+	balancer.Register(clusterImplBB{})
+}
+
+var newXDSClient = func() (xdsClientInterface, error) { return xdsclient.New() }
+
+type clusterImplBB struct{}
+
+func (clusterImplBB) Build(cc balancer.ClientConn, _ balancer.BuildOptions) balancer.Balancer {
+	b := &clusterImplBalancer{
+		ClientConn:     cc,
+		closed:         grpcsync.NewEvent(),
+		loadWrapper:    loadstore.NewLoadStoreWrapper(),
+		pickerUpdateCh: buffer.NewUnbounded(),
+	}
+	b.logger = prefixLogger(b)
+	b.logger.Infof("Created")
+
+	client, err := newXDSClient()
```

Comment:
Nit: It might be better to create the `xdsClient` as the first thing in this function for two reasons:
1. We can assign to `xdsC` as part of the literal struct initialization
2. In the case that the `xdsClient` creation fails, we wouldn't see the log line emitted by `b.logger.Infof("Created")`.

Reply:
We do it in this order for the `logger.Errorf` below. And `logger` needs `b` to print the pointer.

Reply:
Got it. Then maybe, we should move the line `b.logger.Infof("Created")` to be after this `err != nil`, so that we are sure that it is not printed in the case where the xds client creation fails.

Reply:
Done